manpages/imod2vrml.man
.na
.nh
.TH imod2vrml 1 2.00 BL3DEMC
.SH NAME
imod2vrml \- Make a VRML model from an IMOD model file.
.SH SYNOPSIS
imod2vrml [-l]  IMOD_model  VRML_file
.SH DESCRIPTION
Convert an IMOD model file to the VRML 1.0 file format used
by WWW browsers and VRML viewers.  The model will generally 
have similar attributes to
those set in 3dmodv(1) before the conversion.
Specifically, objects that are turned off will not
be included in the VRML file.  For open or closed contour objects,
the various combinations of the "Data Type"
(Contour versus Mesh) and the "Drawing Style" (Points, Lines, Fill, and Fill
Outline) should all appear in the VRML model as they do in 3dmodv.  
However, scattered points will always display as shaded spheres, regardless of
which of these options are selected.
A Fill
Color will be used in place of the line color if that option was set in 3dmodv.
.P
For shaded surfaces, the intensity levels may be quite different
in the VRML viewer than in 3dmodv, depending on the settings in the viewer.
If you are generating files for a particular VRML viewer, you can adjust
the Material
properties in 3dmodv to get the desired appearance in the viewer.  
The Ambient setting will have no effect, so adjust intensity primarily with
the Diffuse setting, and then with the Specular and Shininess settings.
Transparency may also need to be set higher to achieve the desired level, as
well to minimize artifacts from overlapping transparent
surfaces which become more evident at some orientations.
.SH OPTIONS
.TP 
.B -l
Use low-resolution meshes, if any are available.  The default is to use the
highest resolution meshes contained in the model.  Low resolution meshes can
be produced with the 
.B -l
option to imodmesh(1).
.SH AUTHORS
Jim Kremer and David Mastronarde
.SH SEE ALSO
3dmodv(1), imodmesh(1)
.SH BUGS
Email bug reports to mast@colorado.edu.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names
will differ in different Java versions.  This bug will not be fixed because it
does not appear to cause any harm.  Only the automated tests where modified.
(Bug# 1753)


imodutil/imod2vrml.c
/*
 *  imod2vrml - convert IMOD model to VRML version 1
 *
 *  Original author: James Kremer
 *  Revised by: David Mastronarde   email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2005 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id$
 *  Log at end
 */


#include <stdio.h>
#include <string.h>
#include <math.h>
#include "imodel.h"

#define VRML_REDUNDANT 1

static int tablevel;
static int lowres = 0;

/* Function Prototypes */
int imod_to_vrml(Imod *imod, int flags, FILE *fout);
static void printCamera(Imod *imod, FILE *fout);
static void printInfo(Imod *imod, FILE *fout);
static void printLight( Imod *imod, FILE *fout);

static void printObject(Imod *imod, int ob, int flags, FILE *fout);
static void printContours(Imod *imod, Iobj *obj, FILE *fout);
static void printFilledContours(Imod *imod, Iobj *obj, FILE *fout);
static void printScatContours(Imod *imod, Iobj *obj, FILE *fout);
static void printMesh(Imod *imod, int ob, int flags, FILE *fout);

static void usage(int error)
{
  printf("\nConverts an imod model to the Virtual Reality Modeling "
         "Language format.\n");
  printf("Usage: imod2vrml [-l | -s] <imod model file> <output_file.wrl>\n");
  printf("       The -l option selects low-resolution meshes, if any.\n");
  printf("       The -s option outputs separate point/normal list for "
         "each polygon.\n");
  exit(error);
}

int main( int argc, char *argv[])
{
  int i;
  FILE *fout;
  Imod *imod;
  int flags = 0;
  char *progname = imodProgName(argv[0]);

  if (argc < 2) {
    imodVersion(progname);
    imodCopyright();
    usage(0);
  }

  for (i = 1; i < argc; i++){
    if (argv[i][0] == '-')
      switch (argv[i][1]){
      case 'l':
        lowres = 1;
        break;

      case 's':
        flags |= VRML_REDUNDANT;
        break;

      default:
        usage(-1);
        break;

      }
    else
      break;
  }

  if (argc - i != 2)
    usage(-1);

  fout = fopen( argv[i + 1] , "w");
  if (fout == NULL){
    printf("Couldn't open output file %s.\n", argv[i + 1]);
    exit(10);
  }
     
  imod = imodRead(argv[i]);
  if (!imod){
    fprintf(stderr, "%s: Error reading imod model %s\n", argv[0], 
            argv[i]);
    exit(3);
  }
     
  imod_to_vrml(imod, flags, fout);

  fclose(fout);
  exit(0);
}

int imod_to_vrml(Imod *imod, int flags, FILE *fout)
{
  int ob;

  fprintf(fout, "#VRML V1.0 ascii\n");
  fprintf(fout, "\nSeparator {\n");
  {
    printInfo(imod,fout); 
    printLight(imod,fout);
    printCamera(imod,fout);
    for(ob = 0; ob < imod->objsize; ob++)
      printObject(imod, ob,flags, fout);
  }
  fprintf(fout, "}\n");
  return 0;
}


static void printCamera(Imod *imod, FILE *fout)
{
  float xpos, ypos, zpos;
  float xo, yo, zo;
  float fd;
  double r;
  Ipoint maxp, minp;
  float ator = 0.0174533;

  fd = xo = yo = zo = 0.0f;
  imodel_maxpt(imod, &maxp);
  imodel_minpt(imod, &minp);

  maxp.z *= imod->zscale;
  minp.z *= imod->zscale;
  r = ((maxp.x - minp.x) * (maxp.x - minp.x)) +
    ((maxp.y - minp.y) * (maxp.y - minp.y)) +
    ((maxp.z - minp.z) * (maxp.z - minp.z));
  r = sqrt(r);
  xpos = ((maxp.x - minp.x)*0.5)+minp.x;
  ypos = ((maxp.y - minp.y)*0.5)+minp.y;
  zpos = ((maxp.z - minp.z)*0.5)+minp.z;
  fd = r;

  fprintf(fout, "\tDEF Viewer Info { string \"examiner\" }\n");

  fprintf(fout, "\tTranslation { \n\t\ttranslation %g %g %g\n\t}\n",
          -xpos, -ypos, -zpos);

  fprintf(fout, "\tPerspectiveCamera {\n");
  {
    fprintf(fout, "\t\tposition 0 0 %g\n", r);
    fprintf(fout, "\t\torientation %g %g %g 0\n", xo, yo, zo);
    fprintf(fout, "\t\tfocalDistance %g\n", fd);
  }
  fprintf(fout, "\t}\n");
             
  if (imod->cview) {
    fprintf(fout, "\tRotation {\n\t\trotation  1 0 0  %g\n\t}\n",
            ator * imod->view[imod->cview].rot.x);
    fprintf(fout, "\tRotation {\n\t\trotation  0 1 0  %g\n\t}\n",
            ator * imod->view[imod->cview].rot.y);
    fprintf(fout, "\tRotation {\n\t\trotation  0 0 1  %g\n\t}\n",
            ator * imod->view[imod->cview].rot.z);
  }    

  return;
}

static void printLight( Imod *imod, FILE *fout)
{

  fprintf(fout, "\tDirectionalLight {\n"
          "\t\tdirection 1 -1 -1  \n\t}\n");
     
}

static void printInfo(Imod *imod, FILE *fout)
{
  fprintf(fout, "\tSeparator {\n");
  {
    fprintf(fout, "\t\tInfo {\n");
    fprintf(fout, "\t\t\tstring \"Created by Imod\"\n");
    fprintf(fout, "\t\t}\n");
  }
  fprintf(fout, "\t}\n");
  return;
}

static void printMaterial(Iobj *obj, int *lastuse, int usefill, FILE *fout)
{
  float diffuse = ((int)obj->diffuse) / 255.;
  float specular = ((int)obj->specular) / 255.;
  float shininess = ((int)obj->shininess) / 255.;
  float red = obj->red;
  float green = obj->green;
  float blue = obj->blue;
  if (*lastuse == usefill)
    return;

  if (usefill) {
    red = (float)obj->fillred / 255.;
    green = (float)obj->fillgreen / 255.;
    blue = (float)obj->fillblue / 255.;
  }
  *lastuse = usefill;

  /* DMN 1/18/01: VRMLview gave artifacts with shininess at 0 */
  if (shininess < 0.01)
    shininess= 0.01;
     
  /* material definition. */
  fprintf(fout, "\tMaterial{\n");
  /* DNM 1/18/01: in VRMLview ambient color did not affect surface 
     display, did affect line display, so no need to figure out if
     displaying lines or surfaces and scaling by obj->ambient */
  fprintf(fout, "\t\tambientColor %g %g %g\n",
          red, green, blue);
  fprintf(fout, "\t\tdiffuseColor %g %g %g\n",
          red * diffuse, green * diffuse, blue * diffuse);
  fprintf(fout, "\t\tspecularColor %g %g %g\n",
          red * specular, green * specular, blue * specular);
          
  /*      if (!iobjLight(flags))
          fprintf(fout, "\t\temissiveColor %g %g %g\n",
          red, green, blue); */

  /* DNM 1/18/01: this needed to be 0's for proper display */
  fprintf(fout, "\t\temissiveColor 0 0 0\n");
     
  fprintf(fout, "\t\tshininess %g\n", shininess);
  fprintf(fout, "\t\ttransparency %g\n", obj->trans/100.0f);
  fprintf(fout, "\t} #Material\n");
}


static void printObject(Imod *imod, int ob, int flags, FILE *fout)
{
  int co;
  Iobj *obj = &imod->obj[ob];
  int lastuse = -1;
  int hasSpheres = (iobjScat(obj->flags) || obj->pdrawsize) ? 1 : 0;
  if (iobjOff(obj->flags))
    return;

  /* See if object has spheres to draw */
  for (co = 0; co < obj->contsize && !hasSpheres; co++)
    if (obj->cont[co].sizes)
      hasSpheres = 1;

  fprintf(fout, "DEF Object%dData Separator {\n", ob);

  /* Call scattered point routine for all objects with point sizes 
     for some reason this needs to be first */
  if (hasSpheres) {
    printMaterial(obj, &lastuse, obj->flags & IMOD_OBJFLAG_FCOLOR, fout);
    printScatContours(imod, obj, fout);
  }
  
  /* Call routines for drawing various kinds of non-scattered points */
  if (iobjMesh(obj->flags))
    printMesh(imod, ob, flags, fout);
  else if (iobjFill(obj->flags) && iobjClose(obj->flags))
    printFilledContours(imod, obj, fout);
  else if (!iobjScat(obj->flags)) {
    printMaterial(obj, &lastuse, 0, fout);
    printContours(imod, obj, fout);
  }

  fprintf(fout, "} #Object%dData\n", ob);
  return;
}


static void printScatContours(Imod *imod, Iobj *obj, FILE *fout)
{
  Icont *cont;
  int hasSpheres = (iobjScat(obj->flags) || obj->pdrawsize) ? 1 : 0;
  int co, pt;
  float zscale = imod->zscale;
  float size;

  for(co = 0; co < obj->contsize; co++){
    cont = &(obj->cont[co]);
    
    /* Skip the contour if it is empty or if object is not scattered and there 
       are no sizes */
    if (!cont || !cont->psize || (!hasSpheres && !cont->sizes))
      continue;
    for (pt = 0; pt < cont->psize; pt++) {
      size = imodPointGetSize(obj, cont, pt);

      /* Draw point only if size is nonzero */
      if (size > 0.)
        fprintf(fout, "\tDEF PntDat Separator { "
                "Translation { translation %g %g %g}"
                "Sphere { radius %f } }\n",
                cont->pts[pt].x,
                cont->pts[pt].y,
                cont->pts[pt].z * zscale, size);
    }
  }
}

/* Draws open or closed lines if object not filled. */
static void printContours(Imod *imod, Iobj *obj, FILE *fout)
{
  Icont *cont;
  int co, pt;
  float zscale = imod->zscale;

  for(co = 0; co < obj->contsize; co++){
    cont = &(obj->cont[co]);
    if (!cont )
      continue;
    if (!cont->psize)
      continue;

    fprintf(fout, "\tDEF ContourData Coordinate3 {\n");
    fprintf(fout, "\t\tpoint [\n");
    for(pt = 0; pt < cont->psize; pt++)
      fprintf(fout, "\t\t%g %g %g%c\n",
              cont->pts[pt].x,
              cont->pts[pt].y,
              cont->pts[pt].z * zscale,
              (pt == (cont->psize-1))?']':',');
    fprintf(fout, "\t}\n");
    if (iobjLine(obj->flags)) {
      fprintf(fout, "\tIndexedLineSet {\n");
      fprintf(fout, "\t\tcoordIndex [");
      for(pt = 0; pt < cont->psize; pt++) {
        fprintf(fout, "%d,", pt);
        if ((pt % 10) == 9)
          fprintf(fout, "\n");
      }
      if (iobjClose(obj->flags))
        fprintf(fout, "0");
      fprintf(fout, "]\n\t}\n");
    } else {
      fprintf(fout, "\tPointSet {\n\t\tstartIndex 0\n"
              "\t\tnumPoints -1\n\t}\n");
    }            
  }
  return;
}

static void printFilledContours(Imod *imod, Iobj *obj, FILE *fout)
{
  Icont *cont;
  int co, pt;
  float zscale = imod->zscale;
  int lastuse = -1;
     
  for(co = 0; co < obj->contsize; co++){
    cont = &(obj->cont[co]);
    if (!cont )
      continue;
    if (cont->psize < 3)
      continue;

    printMaterial(obj, &lastuse, obj->flags & IMOD_OBJFLAG_FCOLOR, fout);
          
    fprintf(fout, "\tDEF ContourData Coordinate3 {\n");
    fprintf(fout, "\t\tpoint [\n");
    for(pt = 0; pt < cont->psize; pt++)
      fprintf(fout, "\t\t%g %g %g%c\n",
              cont->pts[pt].x,
              cont->pts[pt].y,
              cont->pts[pt].z * zscale,
              (pt == (cont->psize-1))?']':',');
    fprintf(fout, "\t}\n");
    fprintf(fout, "\tIndexedFaceSet {\n");
    fprintf(fout, "\t\tcoordIndex [");
    for(pt = 0; pt < cont->psize; pt++) {
      fprintf(fout, "%d,", pt);
      if (pt % 10 == 9)
        fprintf(fout, "\n");
    }
    if (iobjClose(obj->flags))
      fprintf(fout, "0");
    fprintf(fout, "]\n\t}\n");
    if (iobjLine(obj->flags)) {
      printMaterial(obj, &lastuse, 0, fout);
      fprintf(fout, "\tIndexedLineSet {\n");
      fprintf(fout, "\t\tcoordIndex [");
      for(pt = 0; pt < cont->psize; pt++) {
        fprintf(fout, "%d,", pt);
        if (pt % 10 == 9)
          fprintf(fout, "\n");
      }
      fprintf(fout, "0");
      fprintf(fout, "]\n\t}\n");
    }
  }
}



static void printMesh(Imod *imod, int ob, int flags, FILE *fout)
{
  Iobj *obj = &imod->obj[ob];
  Imesh *mesh;
  int me, i, j;
  float zscale = imod->zscale;
  int *ilist, *mapping;
  int lsize;
  int index, ind, poly;
  Ipoint norm;
  int lastuse = -1;
  int resol;
  int listInc, vertBase, normAdd;

  imodMeshNearestRes(obj->mesh, obj->meshsize, lowres, &resol);

  if (iobjFill(obj->flags))
    fprintf(fout, "\tNormalBinding { value PER_VERTEX }\n");
  else
    fprintf(fout, "\tNormalBinding { value OVERALL }\n");
         
  /* Everywhere below, get rid of extraneous commas and tabs, limit
     resolution to 5 digits for coordinates, 3 digits for normals */
  for(me = 0; me < obj->meshsize; me++){
    if (imeshResol(obj->mesh[me].flag) != resol)
      continue;
    poly = 0;
    mesh = &obj->mesh[me];
    ilist = (int *)malloc(sizeof(int) * mesh->lsize);
    if (!ilist) {
      printf("Error allocating memory\n");
      exit(1);
    }
    if (flags & VRML_REDUNDANT) {
      mapping = (int *)malloc(sizeof(int) * mesh->vsize);
      if (!mapping) {
        printf("Error allocating memory\n");
        exit(1);
      }
    }


    /* Output coordinates unless doing redundant output */
    lsize = 0;
    if (!(flags & VRML_REDUNDANT) || 
        !(iobjFill(obj->flags) || iobjLine(obj->flags))) {
      fprintf(fout, "\tDEF Obj%dMesh%dData  Coordinate3 {\n",
              ob, me);
      fprintf(fout, "\t\tpoint [\n");
      for(i = 0; i < mesh->vsize; i+=2){
        fprintf(fout, "%.5g %.5g %.5g%c\n",
                mesh->vert[i].x,
                mesh->vert[i].y,
                mesh->vert[i].z * zscale,
                (i >= (mesh->vsize-2))?']':',');
      }
      fprintf(fout, "\t}\n");
    }
        
    /* Output normals if doing fill and not doing redundant output */
    if (!(flags & VRML_REDUNDANT) && iobjFill(obj->flags)) {
      fprintf(fout, "\tDEF Obj%dMesh%dNData Normal {\n",
              ob, me);
      fprintf(fout, "\t\tvector [\n");
      for(i = 1; i < mesh->vsize; i+=2){
        norm = mesh->vert[i];
        imodPointNormalize(&norm);
        fprintf(fout, "%.3f %.3f %.3f%c\n",
                norm.x, norm.y, norm.z,
                (i >= (mesh->vsize-1))?']':',');
      }
      fprintf(fout, "\t}\n");
    }

    if (iobjFill(obj->flags) || iobjLine(obj->flags)) {
      for(i = 0; i < mesh->lsize; i++){
        if (mesh->list[i] == IMOD_MESH_END)
          break;
               
        if (imodMeshPolyNormFactors(mesh->list[i], &listInc, &vertBase, 
                                    &normAdd)) {
          
          /* Copy the indices to the new index array numbered from 0 */
          lsize = 0;
          i++;
          while (mesh->list[i] != IMOD_MESH_ENDPOLY) {
            ilist[lsize++] = mesh->list[i + vertBase]/2;
            i += listInc;
            ilist[lsize++] = mesh->list[i + vertBase]/2;
            i += listInc;
            ilist[lsize++] = mesh->list[i + vertBase]/2;
            i += listInc;
          }

          if (!lsize)
            continue;

          /* If doing redundant output, output the points and convert the 
             indexes */
          poly++;
          if (flags & VRML_REDUNDANT) {
            fprintf(fout, "\tDEF Obj%dMesh%dPoly%dData  Coordinate3 {\n",
                    ob, me, poly);
            fprintf(fout, "\t\tpoint [\n");

            index = 0;
            for (j = 0; j < mesh->vsize; j++)
              mapping[j] = -1;
            for (ind = 0; ind < lsize; ind++) {
              j = 2 * ilist[ind];

              /* If mapping doesn't exit yet, output the point */
              if (mapping[j] < 0) {
                mapping[j] = index++;
                if (index > 1)
                  fprintf(fout,",\n");
                fprintf(fout, "%.5g %.5g %.5g",
                        mesh->vert[j].x,
                        mesh->vert[j].y,
                        mesh->vert[j].z * zscale);
              }
            }
            fprintf(fout, "]\n\t}\n");

            /* If doing fill, output normals next, recreate mapping */
            if (iobjFill(obj->flags)) {
              fprintf(fout, "\tDEF Obj%dMesh%dPoly%dNData Normal {\n",
                      ob, me, poly);
              fprintf(fout, "\t\tvector [\n");

              index = 0;
              for (j = 0; j < mesh->vsize; j++)
                mapping[j] = -1;
              for (ind = 0; ind < lsize; ind++) {
                j = 2 * ilist[ind];
                if (mapping[j] < 0) {
                  mapping[j] = index++;
                  if (index > 1)
                    fprintf(fout,",\n");
                  norm = mesh->vert[j + 1];
                  imodPointNormalize(&norm);
                  fprintf(fout, "%.3f %.3f %.3f",
                          norm.x, norm.y, norm.z);
                }
              }
              fprintf(fout, "]\n\t}\n");
            }

            /* Then replace indexes with ones in the polygon-specific set */
            for (ind = 0; ind < lsize; ind++) {
              j = 2 * ilist[ind];
              ilist[ind] = mapping[j];
            }
          }

          /* If showing surface, put out face set */
          if (iobjFill(obj->flags)){
            printMaterial(obj, &lastuse,
                          obj->flags & IMOD_OBJFLAG_FCOLOR, 
                          fout);
            fprintf(fout, "\tIndexedFaceSet {\n");
            fprintf(fout, "\t\tcoordIndex [");
            for(index = 0; index < lsize / 3; index++){
              ind = 3 * index;
              fprintf(fout, "%d,%d,%d,-1%c\n", 
                      ilist[ind],
                      ilist[ind + 1], ilist[ind + 2],
                      (index >= (lsize/3-1))?']':',');
            }
            fprintf(fout,"\t}\n");
          }

          /* If showing lines, put out line set */
          if (iobjLine(obj->flags)){
            printMaterial(obj, &lastuse, 0, fout);
            fprintf(fout, "\tIndexedLineSet {\n");
            fprintf(fout, "\t\tcoordIndex [");
            for(index = 0; index < lsize / 3; index++){
              ind = 3 * index;
              fprintf(fout, "%d,%d,%d,%d,-1%c\n",
                      ilist[ind],
                      ilist[ind + 1], ilist[ind + 2],
                      ilist[ind],
                      (index >= (lsize/3-1))?']':',');
            }
            fprintf(fout, "\t}\n");
          }
                         
        } 
      }
    } else {
      /* If showing neither, define a point set */
      printMaterial(obj, &lastuse, 0, fout);
      fprintf(fout, "\tPointSet {\n\t\tstartIndex 0\n"
              "\t\tnumPoints -1\n\t}\n");
    }
    if (ilist) 
      free(ilist);
    if ((flags & VRML_REDUNDANT) && mapping)
      free(mapping);
  }
}

/*
$Log$
Revision 3.7  2007/04/11 21:22:15  mast
Added option to output points/normals for each polygon

Revision 3.6  2006/08/31 23:13:15  mast
Changed mat1 to real names

Revision 3.5  2005/09/11 19:22:11  mast
Changes for new style of mesh

Revision 3.4  2005/02/03 17:25:09  mast
Prevented incorrect output for empty polygons

Revision 3.3  2005/01/26 22:28:40  mast
Made all points with sizes be displayed as spheres

Revision 3.2  2004/07/07 19:25:30  mast
Changed exit(-1) to exit(3) for Cygwin

Revision 3.1  2002/12/23 21:37:14  mast
fixed exit status

*/
